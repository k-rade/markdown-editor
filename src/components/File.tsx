import { MarkdownFile } from "../types/markdown";
import { FiTrash2 } from "react-icons/fi";
import DeleteFileModal from "./DeleteFileModal";
import { useState } from "react";

interface Props {
  currentFileName: any;
  file: MarkdownFile;
  deleteFile: (name: string) => void;
  selectFile: (name: string) => void;
}

const File = ({
  currentFileName,
  file: { name, markdown },
  deleteFile,
  selectFile,
}: Props): JSX.Element => {
  const [showDeleteFileModal, setShowDeleteFileModal] = useState(false);

  const handleFileDelete = (name: string) => {
    deleteFile(name);
    handleDeleteFileModalToggle();
  };

  const handleDeleteFileModalToggle = (): void => {
    setShowDeleteFileModal((prevState) => !prevState);
  };

  return (
    <div
      className={`${
        currentFileName === name ? "bg-zinc-50 dark:bg-zinc-600" : ""
      } border-b p-4 dark:border-zinc-500`}
    >
      <p className="cursor-pointer mb-1 font-semibold flex items-center gap-2 justify-between dark:text-zinc-200">
        <span onClick={(): void => selectFile(name)}>
          {name ? name : "Untitled"}
        </span>
        <FiTrash2
          data-testid="trash-icon"
          onClick={handleDeleteFileModalToggle}
          size="1.125rem"
          className="stroke-zinc-300 hover:stroke-red-500 hover:cursor-pointer dark:stroke-zinc-400 dark:hover:stroke-red-600"
        />
      </p>
      <p
        onClick={(): void => selectFile(name)}
        className="cursor-pointer overflow-hidden text-ellipsis max-h-16 h-[4rem] text-sm text-zinc-400 dark:text-zinc-400"
      >
        {markdown ? markdown : "Empty note"}
      </p>
      <DeleteFileModal
        isOpen={showDeleteFileModal}
        onClose={handleDeleteFileModalToggle}
        onSubmit={(): void => handleFileDelete(name)}
      />
    </div>
  );
};

export default File;
