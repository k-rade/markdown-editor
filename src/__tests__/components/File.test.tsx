import { fireEvent, render, RenderResult } from "@testing-library/react";
import File from "../../components/File";

describe("File", (): void => {
  const mockFile = { name: "File name", markdown: "markdown text" };

  const mockDeleteFile = jest.fn();
  const mockSelectFile = jest.fn();

  const renderTestComponent = (currentFileName: string): RenderResult =>
    render(
      <File
        currentFileName={currentFileName}
        file={mockFile}
        deleteFile={mockDeleteFile}
        selectFile={mockSelectFile}
      />
    );

  it("should render", (): void => {
    const { baseElement } = renderTestComponent("");

    expect(baseElement).toMatchSnapshot();
  });

  it("should render current select file", (): void => {
    const { baseElement } = renderTestComponent("File name");

    expect(baseElement).toMatchSnapshot();
  });

  it("should delete file", (): void => {
    const { getByTestId, getByText, queryByTestId } = renderTestComponent("");

    fireEvent.click(getByTestId("trash-icon"));
    fireEvent.click(getByText("Delete"));

    expect(mockDeleteFile).toHaveBeenCalledWith("File name");
    expect(queryByTestId("delete-file-modal")).not.toBeInTheDocument();
  });

  it("should select file when clicked on file name", (): void => {
    const { getByText } = renderTestComponent("");

    fireEvent.click(getByText("File name"));

    expect(mockSelectFile).toHaveBeenCalledWith("File name");
  });

  it("should select file when clicked on file markdown", (): void => {
    const { getByText } = renderTestComponent("");

    fireEvent.click(getByText("markdown text"));

    expect(mockSelectFile).toHaveBeenCalledWith("File name");
  });
});
