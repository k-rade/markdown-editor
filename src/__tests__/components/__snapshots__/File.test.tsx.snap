// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File should render 1`] = `
<body>
  <div>
    <div
      class=" border-b p-4 dark:border-zinc-700"
    >
      <p
        class="cursor-pointer mb-1 font-semibold flex items-center gap-2 justify-between dark:text-zinc-200"
      >
        <span>
          File name
        </span>
        <svg
          class="stroke-zinc-300 hover:stroke-rose-500 hover:cursor-pointer dark:stroke-zinc-500 dark:hover:stroke-rose-800"
          data-testid="trash-icon"
          fill="none"
          height="1.125rem"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1.125rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="3 6 5 6 21 6"
          />
          <path
            d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
          />
          <line
            x1="10"
            x2="10"
            y1="11"
            y2="17"
          />
          <line
            x1="14"
            x2="14"
            y1="11"
            y2="17"
          />
        </svg>
      </p>
      <p
        class="cursor-pointer overflow-hidden text-ellipsis max-h-16 h-[4rem] text-sm text-zinc-400 dark:text-zinc-400"
      >
        markdown text
      </p>
    </div>
  </div>
</body>
`;

exports[`File should render current select file 1`] = `
<body>
  <div>
    <div
      class="bg-zinc-50 dark:bg-zinc-700/25 border-b p-4 dark:border-zinc-700"
    >
      <p
        class="cursor-pointer mb-1 font-semibold flex items-center gap-2 justify-between dark:text-zinc-200"
      >
        <span>
          File name
        </span>
        <svg
          class="stroke-zinc-300 hover:stroke-rose-500 hover:cursor-pointer dark:stroke-zinc-500 dark:hover:stroke-rose-800"
          data-testid="trash-icon"
          fill="none"
          height="1.125rem"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="1.125rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="3 6 5 6 21 6"
          />
          <path
            d="M19 6v14a2 2 0 0 1-2 2H7a2 2 0 0 1-2-2V6m3 0V4a2 2 0 0 1 2-2h4a2 2 0 0 1 2 2v2"
          />
          <line
            x1="10"
            x2="10"
            y1="11"
            y2="17"
          />
          <line
            x1="14"
            x2="14"
            y1="11"
            y2="17"
          />
        </svg>
      </p>
      <p
        class="cursor-pointer overflow-hidden text-ellipsis max-h-16 h-[4rem] text-sm text-zinc-400 dark:text-zinc-400"
      >
        markdown text
      </p>
    </div>
  </div>
</body>
`;
